common/qeidentity.c

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef OE_USE_LIBSGX
+#include "qeidentity.h"
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/bits/safemath.h>
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/cert.h>
+#include <openenclave/internal/crl.h>
+#include <openenclave/internal/datetime.h>
+#include <openenclave/internal/ec.h>
+#include <openenclave/internal/hexdump.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/report.h>
+#include <openenclave/internal/sgxcertextensions.h>
+#include <openenclave/internal/sha.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+#include "common.h"
+#include "tcbinfo.h"
+
+// hardcoded property values used for validating quoting enclave when qe
+// identity info is not available
+// The mrsigner value of Intel's Production quoting enclave.
+static const uint8_t g_qe_mrsigner[32] = {
+    0x8c, 0x4f, 0x57, 0x75, 0xd7, 0x96, 0x50, 0x3e, 0x96, 0x13, 0x7f,
+    0x77, 0xc6, 0x8a, 0x82, 0x9a, 0x00, 0x56, 0xac, 0x8d, 0xed, 0x70,
+    0x14, 0x0b, 0x08, 0x1b, 0x09, 0x44, 0x90, 0xc5, 0x7b, 0xff};
+
+// The isvprodid value of Intel's Production quoting enclave.
+static const uint32_t g_qe_isvprodid = 1;
+
+// The isvsvn value of Intel's Production quoting enclave.
+static const uint32_t g_qeisvsvn = 1;
+
+void dump_info(char* title, uint8_t* data, uint8_t count)
+{
+    OE_TRACE_INFO("%s\n", title);
+    for (uint8_t i = 0; i < count; i++)
+    {
+        OE_TRACE_INFO("[%d] = %x\n", i, data[i]);
+    }
+}
+
+oe_result_t oe_enforce_qe_identity(sgx_report_body_t* qe_report_body)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_get_qe_identity_info_args_t qe_id_args = {0};
+    const uint8_t* pem_pck_certificate = NULL;
+    size_t pem_pck_certificate_size = 0;
+    oe_cert_chain_t pck_cert_chain = {0};
+    oe_parsed_qe_identity_info_t parsed_info = {0};
+
+    OE_TRACE_INFO("Calling %s\n", __PRETTY_FUNCTION__);
+
+    // fetch qe identity information
+    result = oe_get_qe_identity_info(&qe_id_args);
+    if (result == OE_QUOTE_PROVIDER_CALL_ERROR)
```

Comment:
This is good for the time being. But long term we need to error out if QE identity is not available.

Reply:
Agree.

Reply:
I will add some note on this.